py/desiutil/timer.py

```diff
@@ -0,0 +1,125 @@
+"""
+timer: standardizing reporting of algorithm and I/O timing
+
+The `timer.Timer` class is intended for reporting timing of events that
+take seconds to minutes; it is not intended for detailed profiling.  Example::
+
+    from desiutil.timer import Timer
+    t = Timer()
+    t.start('blat')
+
+    #- Use context manager for timing simple steps, e.g. one-liners
+    with t.time('blat.input'):
+        stuff = io.read(filename)
+
+    #- Or use full start/stop if you prefer
+    t.start('blat.algorithm')
+    results = calculate(stuff)
+    t.stop('blat.algorithm')
+
+    with t.time('blat.output'):
+        io.write(results)
+
+    #- Stop outer timer; Note that named timers can be nested or interleaved
+    t.stop('blat')
+
+    #- Print a json report of the timing
+    print(t.report())
+"""
+
+import traceback
+import time
+import datetime
+import json
+import copy
+import os.path
+from contextlib import contextmanager
+
+class Timer(object):
+    """
+    A basic timer class for standardizing reporting of algorithm and I/O timing
+    
+    TIMER:<START|STOP>:<filename>:<lineno>:<funcname>: <message>
+    """
+    
+    def __init__(self):
+        self.timers = dict()
+    
+    def _prefix(self, step):
+        """
+        Return standard prefix string for timer reporting
+        
+        Args:
+            step (str): timing step, e.g. "START" or "STOP"
+        """
+        stack = traceback.extract_stack()
+        #- Walk backwards in stack to find first caller not from this file
+        #- and not contextlib.py
+        #- (exact index depends on whether context manager was used or not)
+        thisfile = os.path.normpath(__file__)
+        for caller in stack[-1::-1]:
+            if os.path.normpath(caller.filename) != thisfile and \
+               os.path.basename(caller.filename) != 'contextlib.py':
+                break
+        
+        filename = os.path.basename(caller.filename)
+        return f"TIMER-{step}:{filename}:{caller.lineno}:{caller.name}:"
+    
+    def start(self, name):
+        """Start a timer `name` (str)"""
+        # timestamp = datetime.datetime.now().replace(microsecond=0).isoformat()        
+        timestamp = datetime.datetime.now().isoformat()        
+        if name in self.timers:
+            print(self._prefix('WARNING'), f'Restarting {name} at {timestamp}')
+        
+        print(self._prefix('START'), f'Starting {name} at {timestamp}')
+        self.timers[name] = dict(start=time.time())
+    
+    def stop(self, name):
+        """Stop timer `name` (str)"""
+        #- non-fatal ERROR: trying to stop a timer that wasn't started
+        timestamp = datetime.datetime.now().isoformat()        
+        if name not in self.timers:
+            print(self._prefix('ERROR'), f'Tried to stop non-existent timer {name} at {timestamp}')
```

Comment:
I think it would be better to raise an exception here.  If that is not possible, at least return a `TimeDelta` rather than some other type.

Reply:
I documented the philosophy more, but I do want to stick with the safety that adding timer info shouldn't be able to crash user code, even if used incorrectly.  This is an exception to my normal guideline of fail-early fail-often.

Also, internally the timestamps are tracked as `time.time()` seconds since epoch, not `datetime` instances, so a correct call to `stop` returns a float difference in seconds not a `timedelta`, so I kept the incorrect call to stop consistent with that by returning a float.  I updated the documentation to be clear about what is returned.